### ISSUE TITLE
 Avoid inline-styling to meet tighter content-security-policy

### ISSUE THREAD

**Issue by KKoukiou** [patternfly/patternfly-react#2919]
@priley86 This was reported first here - https://github.com/patternfly/patternfly-react/issues/2606
And this was the fix PR https://github.com/patternfly/patternfly-react/pull/2653

But using the versions that are supposed to have this fix:
-    "@patternfly/react-core": "3.86.2",
-    "@patternfly/react-table": "2.16.16

Don't seem to remove the inline style warnings.

Maybe there are still some remainders?
```
  | makeStyleTag | @ | index.esm.js:145
-- | -- | -- | --
  | inject | @ | index.esm.js:167
  | createEmotion | @ | index.esm.js:265
  | (anonymous) | @ | index.esm.js:5
  | ./node_modules/emotion/dist/index.esm.js | @ | index.esm.js:17
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-styles/dist/esm/StyleSheet.js | @ | StyleSheet.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-styles/dist/esm/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-core/dist/esm/components/AboutModal/AboutModal.js | @ | AboutModal.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-core/dist/esm/components/AboutModal/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-core/dist/esm/components/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-core/dist/esm/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-table/dist/esm/components/Table/ActionsColumn.js | @ | ActionsColumn.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-table/dist/esm/components/Table/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-table/dist/esm/components/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./node_modules/@patternfly/react-table/dist/esm/index.js | @ | index.js:1
  | __webpack_require__ | @ | bootstrap:19
  | ./pkg/machines/components/vmDisksTab.jsx | @ | vmDisksTab.jsx:1
  | __webpack_require__ | @ | bootstrap:19
  | ./pkg/machines/components/vmDisksTabLibvirt.jsx | @ | vmDisksTabLibvirt.jsx:1
  | __webpack_require__


```

**Comment by priley86** [comment]
Hey @KKoukiou - can you provide any further insight here? Our fix attempted to remove all explicit references of `Stylesheet.parse` and `Stylesheet.inject`, however that does not seem to be agreeing w/ your CSP checker.

This kind of takes me back to my original thoughts here and looking at the `nonce` option:
https://github.com/patternfly/patternfly-react/issues/2606#issuecomment-516063260

I am not sure how easy it will be to pull out the `makeStyleTag` bit, but maybe we can pass a `nonce` to it to avoid this warning you are receiving...


thoughts @dgutride ?

**Comment by KKoukiou** [comment]
Hey @priley86 I am not sure how I can help you solve this - I can help you test any changes before merging though against cockpit.

So I see you use this module, https://github.com/emotion-js/emotion,
which is where makeStyleTag that our webpack is complaining about is taken from.
What is this doing that you can't do by importing CSS and styling the components with classes?
I will be glad if you can give me some heads up about the need of this module in the PF-react code :)

**Comment by dgutride** [comment]
@KKoukiou  - our current method of including css should work with what you are asking - can you create a simple repo with a repro case for this issue so we can test it out?  @redallen @priley86 ...

**Comment by KKoukiou** [comment]
> @KKoukiou - our current method of including css should work with what you are asking - can you create a simple repo with a repro case for this issue so we can test it out? @redallen @priley86 ...

Sure, here is a minimal reproducer https://github.com/KKoukiou/patternfly-react-table-example

Clone the repo, install the npm dependencies with 'npm install .' run webpack to bundle the code with 'webpack --progress' command and then run 'npm run-script start' to serve the page with webpack-dev-server.

In the browser console you can see the same warnings that we are getting in cockpit when using the Table component.

**Comment by priley86** [comment]
so one possibility here is modifying our usage of Emotion to add nonce's as mentioned previously. There is various ways of achieving this, but essentially just adding nonce keys to your style tags... we could add support for a global using `create-emotion` API like so:
https://github.com/priley86/patternfly-react/commit/a6bd0ede2d5bc8ddd3d77251a7cc718d02c0f211

`create-emotion` docs: https://emotion.sh/docs/create-emotion#primary-use-cases

And consuming this  would be like this (haven't figured out why miniCSSExtract does not respect nonce's yet, but style-loader seems to respect them):
https://github.com/priley86/patternfly-react-table-example/commit/d713044d67fa1211319cdac0849b1f1befd84594

The `CacheProvider` bit was just me testing Emotion's new Cache API which does this an alternative way using a Context...

`@emotion/core` and `@emotion/cache` are here:
https://emotion.sh/docs/@emotion/core
https://emotion.sh/docs/@emotion/cache

This method may be a little more involved (and overkill if we intend to drop Emotion any way). 😸 

Just some ideas that would help you resolve this though...

**Comment by dgutride** [comment]
@KKoukiou  - quick update from our end.  Thank you very much for the repo with info on how to reproduce this.  The use of emotion (just being required by an import statement) is causing this, none of our code.  For now, I'd like to remove emotion as a required import by our normal use cases, but it's going to take a bit of work.  Can you give us an idea of when you'd need to have this by?

**Comment by KKoukiou** [comment]
> @KKoukiou - quick update from our end. Thank you very much for the repo with info on how to reproduce this. The use of emotion (just being required by an import statement) is causing this, none of our code. For now, I'd like to remove emotion as a required import by our normal use cases, but it's going to take a bit of work. Can you give us an idea of when you'd need to have this by?

Since the blocked css does not affect somehow the look of the used components I don't think this is extra urgent. We will just keep using the affected components and we will just see the warnings for the suppressed styles.

**Comment by KKoukiou** [comment]
So unfortunately I hit the first blocking issue for us related to this:

react-charts are unusable without inlines styles.

The colors are not working, the tooltips neither.
The documentation example looks basically like this:
![Screenshot_2019-10-21 React and Webpack4](https://user-images.githubusercontent.com/14921356/67195400-297d6200-f3f9-11e9-88eb-d1605af6c8ac.png)

So at least for the charts I would appreciate a fix asap so that we can start using them.

**Comment by rachael-phillips** [comment]
Hi @KKoukiou I've seen this and I will be looking at this with the team first thing tomorrow to determine next steps.

**Comment by dgutride** [comment]
@KKoukiou  - we won't be able to provide a quick fix unfortunately since it's a rather invasive change we need to make to avoid pulling in emotion to our libraries.  We'd have to investigate changing how styles are consumed in react-charts and the overall fix we'd recommend will be fairly involved.  Do you know when you are planning to ship your first release with charts and how long we have before your code must be implemented?

**Comment by KKoukiou** [comment]
We are currently using https://www.npmjs.com/package/c3 for charts, moving to PF4 charts would be an enhancement for us so we don't have strict deadlines for that.

**Comment by dgutride** [comment]
@KKoukiou  - I'm going to have to move the react-styles change to a breaking change fix which could be scheduled for .a later time.  I did try this out with your example and added a donut chart from react-charts to see if it was broken for me.  You will need to change the order of the imports to make sure the base.css is loaded first to get the styling correct:

```
import "@patternfly/react-core/dist/styles/base.css";
import { App } from './App';
```

Please try that out and let me know - once we correct the charts issue for you - I'm going to mark this as a breaking change issue (with the fix completely removing emotion from patternfly-react across the board).

**Comment by redallen** [comment]
Should be closed by #3886. Here are some prereleases:

```
 - @patternfly/react-catalog-view-extension@4.0.5
 - @patternfly/react-charts@6.0.3
 - @patternfly/react-core@4.0.6
 - @patternfly/react-inline-edit-extension@4.0.6
 - @patternfly/react-styles@4.0.3
 - @patternfly/react-table@4.0.6
 - @patternfly/react-topology@4.0.6
 - @patternfly/react-virtualized-extension@4.0.6
```

**Comment by redallen** [comment]
I just tested this in our react seed app with the following versions:
```
    "@patternfly/react-core": "^4.0.6",
    "@patternfly/react-icons": "^4.0.3",
    "@patternfly/react-styles": "^4.0.3",
```
I'm happy to report there are no `<style>` tags in the `<head>`!
![image](https://user-images.githubusercontent.com/47335686/76993967-15944880-6924-11ea-83cf-18888c307420.png)
